### ISSUE TITLE
About norm

### ISSUE THREAD

**Issue by SongFGH** [tkipf/gae#17]
Hi @tkipf:
I have question about norm in train.py [https://github.com/tkipf/gae/blob/master/gae/train.py](url).
So ,can you explain it?

**Comment by tkipf** [comment]
Yes, have a look here:
https://arxiv.org/abs/1609.02907

**Comment by SongFGH** [comment]
Thanks!

**Comment by Tomposon** [comment]
Do you understand norm now? @SongFGH

**Comment by SongFGH** [comment]
actually, it means we should give more attention to non-zero element in adjacent matrix

**Comment by Tomposon** [comment]
Oh, thank you very much!

**Comment by Tomposon** [comment]
But I find it use norm to multiply the final loss, not only to enlarge the non-zeros element loss like in SDNE.  Did you exactly find the answer from the paper the author recommend to you?

self.cost = norm * tf.reduce_mean(tf.nn.weighted_cross_entropy_with_logits(logits=preds_sub, targets=labels_sub, pos_weight=pos_weight)) @SongFGH

**Comment by SongFGH** [comment]
I don't find the answer from the paper recommended.
All I said is my guess!
Supposeweight of zeros element loss is 1, weight of non-zeros element is norm .

**Comment by hylBen** [comment]
Well, there are (n_nodes * n_nodes) inter-node relationships in total, some of them are edges while others are not. Suppose we have RE edges and RN non-edges, such that

1. RE + RN = (n_nodes * n_nodes),
2. RN / RE = K.

The intuition behind *norm*, I guess, is to **augment edges by replicating edges and corresponding predictions by K times**, so that edges and the non-edge counterparts are balanced in terms of quantity. Of course, you don't have to manually do that since the argument _pos_weight_ in _tf.nn.weighted_cross_entropy_with_logits_ has attained an equivalent result.

Now we not only balanced training data but also increment the size from (n_nodes * n_nodes) to (K*RE + RN) = (2 * RN). However, when applying _tf.reduce_mean_ to the inner cross-entropy, the latter is only divided by  (n_nodes * n_nodes), that's why we need to renormalize it by multiplying a term 

norm = "size of original training set" / "size of augmented training size" = (n_nodes * n_nodes) / (2 * RN)

**Comment by udayshankars** [comment]
@hylBen Thank you so much for the explanation!
